### PR TITLE
Refactor LLMs

### PR DESCRIPTION
**Pull request by maykcaldas**

### CODE REVIEW

File: llmclient/llms.py [hunk 1]
```diff
@@ -143,24 +142,26 @@ def validate_json_completion(
...
 async def do_callbacks(
     async_callbacks: Iterable[Callable[..., Awaitable]],
     sync_callbacks: Iterable[Callable[..., Any]],
-    chunk: str,
+    completion: str,
     name: str | None,
 ) -> None:
     for f in async_callbacks:
-        args, kwargs = prepare_args(f, chunk, name)
+        args, kwargs = prepare_args(f, completion, name)
```
Comment (jamesbraza) [inline llmclient/llms.py:161]: Can we `asyncio.gather` the async callbacks

File: llmclient/llms.py [hunk 2]
```diff
@@ -279,78 +327,145 @@ async def run_prompt(
                     else [human_message_prompt]
                 )
             ]
-            return await self._run_chat(messages, callbacks, name, system_prompt)
+            results = await self._run_chat(messages, callbacks, name)
```
Comment (jamesbraza) [inline llmclient/llms.py:330]: We should not support `run_prompt` and `call`, they seem like the same thing, no?

I would prefer we have "one opinionated way" of using `LLMModel`

File: llmclient/llms.py [hunk 3]
```diff
@@ -279,78 +327,145 @@ async def run_prompt(
...
+            if not sum(usage):
+                result.completion_count = self.count_tokens(result.text)
+            result.seconds_to_last_token = (
+                asyncio.get_running_loop().time() - start_clock
+            )
+            if self.llm_result_callback:
+                if is_coroutine_callable(self.llm_result_callback):
+                    await self.llm_result_callback(result)  # type: ignore[misc]
+                else:
+                    self.llm_result_callback(result)
```
Comment (jamesbraza) [inline llmclient/llms.py:467]: Okay several comments here (all starting from the `type: ignore[misc]`):
- `llm_result_callback`: type hint return with `Any` instead of `None`. We don't care about the return type since we throw it away, so we don't need to require it be `None`
- `llm_result_callback`: type hint as `Callable[[LLMResult], Any | Awaitable[Any]]`, and then apply the same fix from https://github.com/Future-House/ldp/pull/213 here?

Then this code becomes:

```python
            if self.llm_result_callback:
            possibly_awaitable_result = self.llm_result_callback(result)
            if inspect.isawaitable(possibly_awaitable_result):
                await possibly_awaitable_result
```

File: llmclient/llms.py (hunk 2 above)
Comment (jamesbraza) [reply llmclient/llms.py:330]: Oh let's drop `run_prompt` and move PaperQA to whatever our unified interface will be

The point of this refactor is to be a big breaking change to get us to our "ideal" place

Comment (maykcaldas) [reply llmclient/llms.py]: Dropped it. Now we only support `LLMModel.call` and the input is a list of `aviary.Message`s

File: llmclient/llms.py (hunk 1 above)
Comment (maykcaldas) [reply llmclient/llms.py]: Change it to

``` python
await asyncio.gather(
        *(
            f(*args, **kwargs)
            for f in async_callbacks
            for args, kwargs in [prepare_args(f, completion, name)]
        )
    )
```

File: llmclient/llms.py (hunk 2 above)
Comment (jamesbraza) [reply llmclient/llms.py]: Thank you!

File: llmclient/llms.py [hunk 4]
```diff
@@ -169,37 +174,37 @@ class LLMModel(ABC, BaseModel):
...
             "An async callback that will be executed on each"
-            " LLMResult (different than callbacks that execute on each chunk)"
+            " LLMResult (different than callbacks that execute on each completion)"
         ),
         exclude=True,
     )
     config: dict = Field(default_factory=dict)
 
-    async def acomplete(self, prompt: str) -> Chunk:
+    async def acomplete(self, prompt: str) -> LLMResult:
```
[the remark this answers is not in the document]
Comment (maykcaldas) [reply llmclient/llms.py:187]: `acomplete` and `achat` are used respectively for `completion` and `chat` models. `call` checks the llm type and call the correct method accordingly. 

Maybe collapsing `achat` with `_run_chat` and `acomplete` with `_run_complation` makes more sense?

[the remark this answers is not in the document]
Comment (jamesbraza) [reply llmclient/llms.py:187]: Yeah the point is we want to have just less possible control flows, and looking at the type signatures left a few methods are redundant.

Then the ones we keep, we should ideally put a little blurb in the `README` documenting the rationale behind our chosen interface

File: llmclient/llms.py [hunk 5]
```diff
@@ -384,55 +470,62 @@ async def _run_completion(
...
+                await possibly_awaitable_result
         return result
 
 
 LLMModelOrChild = TypeVar("LLMModelOrChild", bound=LLMModel)
 
 
 def rate_limited(
-    func: Callable[[LLMModelOrChild, Any], Awaitable[Chunk] | AsyncIterable[Chunk]],
+    func: Callable[
```
Comment (jamesbraza) [inline llmclient/llms.py:510]: Are we able to streamline this type signature at all?

I am wondering if we can resolve https://github.com/Future-House/llm-client/issues/33 here

Comment (maykcaldas) [reply llmclient/llms.py:370]: Can you show me how to add `--doctest-modules` or check the `typeguard` so I can see the error we get and debug it?

Comment (jamesbraza) [reply llmclient/llms.py:370]: In `tool.pytest.ini_options` you add `addopts = "--doctest-modules --typeguard-packages=lmi"`

Comment (maykcaldas) [reply llmclient/llms.py:371]: I added typeguard and changed `rate_limited` typing to pass the tests. This includes the types for `acompletion`, `acompletion_iter`, and the ones used in `cost_tracker`. However, I still needed to cast `func` inside `rate_limited`  because `mypy` was not accepting that `isasyncgenfunction(func)` would catch the `AsyncIterator` part of `func`

File: tests/test_llms.py [hunk 6]
```diff
@@ -68,17 +67,24 @@ async def test_call(self, config: dict[str, Any]) -> None:
...
+        assert isinstance(result.prompt[1], Message)
+        assert all(isinstance(msg, Message) for msg in result.prompt)
+        assert len(result.prompt) == 2  # role + user messages
+        assert result.prompt[1].content
+        assert result.text
+
+        result = await llm.call_single(messages)
+        assert isinstance(result, LLMResult)
+
+    # @pytest.mark.vcr(match_on=[*VCR_DEFAULT_MATCH_ON])
```
Comment (jamesbraza) [inline tests/test_llms.py:84]: Why comment this? Can you un-comment it?

File: llmclient/llms.py [hunk 7]
```diff
@@ -484,16 +429,16 @@ class PassThroughRouter(litellm.Router):  # TODO: add rate_limited
     def __init__(self, **kwargs):
         self._default_kwargs = kwargs
 
-    async def atext_completion(self, *args, **kwargs):
-        return await litellm.atext_completion(*args, **(self._default_kwargs | kwargs))
```
Comment (jamesbraza) [inline llmclient/llms.py:488]: Let's keep this around since `PassThroughRouter` is different from `LLMModel`.

I think we still want the `PassThroughRouter` to wrap both methods

File: tests/test_llms.py (hunk 6 above)
Comment (maykcaldas) [reply tests/test_llms.py:84]: Specifically in the case of  images, I saw the `body` changing  and then vcr tried to save a new cassette. Which kept giving me this error: `raise CannotOverwriteExistingCassetteException(cassette=cassette, failed_request=vcr_request)`. Even if I removed the match on the body, the test was not passing

File: tests/test_llms.py [hunk 8]
```diff
@@ -28,35 +27,35 @@ class TestLiteLLMModel:
...
                         }
                     ],
                 },
-                id="chat-model",
+                id="OpenAI-model",
             ),
             pytest.param(
                 {
-                    "model_name": "gpt-3.5-turbo-instruct",
+                    "model_name": CommonLLMNames.ANTHROPIC_TEST.value,
```
Comment (jamesbraza) [inline tests/test_llms.py:46]: So going from `gpt-3.5-turbo-instruct` to Anthropic is not a 1-1 adaptation, you're actually changing the test here.

I vote we remove this case, as opposed to going to Anthropic

File: llmclient/llms.py (hunk 7 above)
Comment (maykcaldas) [reply llmclient/llms.py:488]: Even thought we don't use `atext_completion` anymore, does it make sense to have it in the by-pass?

File: tests/test_llms.py (hunk 6 above)
Comment (jamesbraza) [reply tests/test_llms.py:84]: Update: I realize you didn't change this, just GitHub diff makes it look this way

File: llmclient/llms.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (maykcaldas) [reply llmclient/llms.py]: Just for the record: I dropped support to completion models (`atext_completion` endpoint from `litellm`) and renamed our `achat` methods to `acompletion` to make it similar to `litellm`'s nomenclature

File: tests/test_llms.py [hunk 9]
```diff
@@ -369,12 +400,20 @@ def play(move: int | None) -> None:
     @pytest.mark.parametrize(
         ("model_name", "output_type"),
         [
-            pytest.param("gpt-3.5-turbo", DummyOutputSchema, id="json-mode-base-model"),
```
Comment (jamesbraza) [inline tests/test_llms.py:372]: We should bring back `gpt-3.5-turbo` because it actually has different properties than `gpt-4o`

In general, when switching to `CommonLLMNames`, let's make sure it's not changing crafted edge case inputs

File: tests/test_llms.py (hunk 8 above)
Comment (maykcaldas) [reply tests/test_llms.py:46]: I changed it because since we dropped the support to `atext_completion`, we couldn't run `gpt-3.5-turbo-instruct` anymore. Then I noticed that antrophic models didn't have much coverage and decided to use it instead of just removing the test

File: tests/test_llms.py (hunk 6 above)
Comment (maykcaldas) [reply tests/test_llms.py:84]: I think this was changed a while ago when we implemented the `call` interface and tested `Message`s with images.  But it still annoys me that we are not using `vcr` for t his test

File: llmclient/llms.py (hunk 7 above)
Comment (jamesbraza) [reply llmclient/llms.py:488]: So this PR is related to core `LLMModel` refactors, imo it doesn't need to also change this helper adapter entity.

I think we keep it in-tact for this PR

File: tests/test_llms.py (hunk 8 above)
Comment (jamesbraza) [reply tests/test_llms.py:46]: Okay gotchu, sure we can keep this case around then.

Tho I prefer we don't rename the other case

Comment (maykcaldas) [reply tests/test_llms.py:46]: I will create a new `CommonLLMNames` attribute to have a `gpt-3.5` name. Then I'll keep the same models for the tests. Does it sounds good?

Comment (maykcaldas) [reply tests/test_llms.py:46]: Yeah, I to make it easier in case we wanted to update tests for a different model in the future. 

Updated with a new `GPT_35` entry, but if you think it's better to focus on `LLModel` changes in this PR and make this change in a different PR, I'm also down for that.

File: llmclient/llms.py [hunk 10]
```diff
@@ -73,6 +71,7 @@ class CommonLLMNames(StrEnum):
 
     # Use these to avoid thinking about exact versions
     GPT_4O = "gpt-4o-2024-11-20"
+    GPT_35 = "gpt-3.5-turbo-0125"
```
Comment (jamesbraza) [inline llmclient/llms.py:74]:
```suggestion
    GPT_35_TURBO = "gpt-3.5-turbo-0125"
```

I think there was a GPT 3.5 before GPT 3.5 turbo, so it'd be good to disambiguate

File: llmclient/llms.py [hunk 11]
```diff
@@ -586,76 +549,82 @@ async def check_rate_limit(self, token_count: float, **kwargs) -> None:
...
-            completion_tokens=response.usage.completion_tokens,  # type: ignore[attr-defined]
-        )
-
-    @rate_limited
-    async def achat_iter(  # type: ignore[override]
-        self, messages: list[Message]
-    ) -> AsyncIterable[Chunk]:
+    async def acompletion_iter(  # type: ignore[override]
+        self, messages: list[Message], **kwargs
+    ) -> AsyncGenerator[LLMResult]:
```
Comment (jamesbraza) [inline llmclient/llms.py:591]:
```suggestion
    ) -> AsyncIterator[LLMResult]:
```

Base class has `AsyncIterator[LLMResult]`, why does this have `AsyncGenerator[LLMResult]`?

File: llmclient/llms.py [hunk 12]
```diff
@@ -586,76 +549,82 @@ async def check_rate_limit(self, token_count: float, **kwargs) -> None:
...
-            text=cast(litellm.Choices, response.choices[0]).message.content,
-            prompt_tokens=response.usage.prompt_tokens,  # type: ignore[attr-defined]
-            completion_tokens=response.usage.completion_tokens,  # type: ignore[attr-defined]
-        )
-
-    @rate_limited
-    async def achat_iter(  # type: ignore[override]
-        self, messages: list[Message]
-    ) -> AsyncIterable[Chunk]:
+    async def acompletion_iter(  # type: ignore[override]
```
Comment (jamesbraza) [inline llmclient/llms.py:589]: Are we able to resolve the `type: ignore[override]`?

File: llmclient/llms.py [hunk 13]
```diff
@@ -586,76 +549,82 @@ async def check_rate_limit(self, token_count: float, **kwargs) -> None:
...
+            role = delta.role or role
+
+        text = "".join(outputs)
+        result = LLMResult(
+            model=self.name,
+            text=text,
+            prompt=messages,
+            messages=[Message(role=role, content=text)],
+            # TODO: Can we marginalize over all choices?
+            # logprob=sum_logprobs(completion),
```
Comment (jamesbraza) [inline llmclient/llms.py:616]: Are we able to resolve this `TODO`?

File: llmclient/llms.py [hunk 14]
```diff
@@ -586,76 +549,82 @@ async def check_rate_limit(self, token_count: float, **kwargs) -> None:
...
     @rate_limited
-    async def acomplete(self, prompt: str) -> Chunk:  # type: ignore[override]
-        response = await track_costs(self.router.atext_completion)(
-            model=self.name, prompt=prompt
-        )
-        return Chunk(
-            text=response.choices[0].text,
-            prompt_tokens=response.usage.prompt_tokens,
-            completion_tokens=response.usage.completion_tokens,
+    async def acompletion(self, messages: list[Message], **kwargs) -> list[LLMResult]:  # type: ignore[override]
```
Comment (jamesbraza) [inline llmclient/llms.py:552]: Can we somehow resolve this `type: ignore[override]`?

**Review by jamesbraza — COMMENTED** [review]
Getting closer

File: llmclient/llms.py (hunk 10 above)
Comment (maykcaldas) [reply llmclient/llms.py:74]: If I remember correctly, they called they had gpt-3 family that were called by ada, babbage, curie, and davinci. Then the gpt-3.5 appeared after them. But yeah, some people called davinci the "gpt-3.5". It's good to use `GPT_35_TURBO` to make it clear

File: llmclient/llms.py (hunk 11 above)
Comment (maykcaldas) [reply llmclient/llms.py:591]: Oh, nice catch. Both were `AsyncIterator` before. But while changing types in `rate_limited` I noticed that it is just `yield`ing the `LLMResults`. So I think this is actually an `AsyncGenerator`. Forgot to adjust the base class. What do you think? Is it correct that `acompletion_iter` should return an `AsyncGenerator`?

File: llmclient/llms.py (hunk 12 above)
Comment (maykcaldas) [reply llmclient/llms.py:589]: Maybe the case is I couldn't find how to solve it, but because `rate_limited` changes the output from `LiteLLMModel.acompletion`, `mypy` thinks its signature is incompatible with `LLMModel.acompletion`'s signature.

File: llmclient/llms.py (hunk 13 above)
Comment (maykcaldas) [reply llmclient/llms.py]: Implemented it
